packages/bundle-source/cache.js

```diff
@@ -0,0 +1,346 @@
+   */
+  const validateOrAdd = async (rootPath, targetName, log = defaultLog) => {
+    let meta;
+    if (wr.readOnly().neighbor(toBundleMeta(targetName)).exists()) {
+      try {
+        meta = await validate(targetName, rootPath, log);
+        const { bundleTime, contents } = meta;
+        log(
+          `${wr}`,
+          toBundleName(targetName),
+          'valid:',
+          contents.length,
+          'files bundled at',
+          bundleTime,
+        );
+      } catch (invalid) {
+        log(invalid);
+      }
+    }
+    if (!meta) {
+      log(`${wr}`, 'add:', targetName, 'from', rootPath);
+      meta = await add(rootPath, targetName, log);
+      const { bundleFileName, bundleTime, contents } = meta;
+      log(
+        `${wr}`,
+        'bundled',
+        contents.length,
+        'files in',
+        bundleFileName,
+        'at',
+        bundleTime,
+      );
+    }
+    return meta;
+  };
+
+  const loaded = new Map();
+  /**
+   * @param {string} rootPath
+   * @param {string} [targetName]
+   * @param {(...args: any[]) => void} [log]
+   */
+  const load = async (
+    rootPath,
+    targetName = readPowers.basename(rootPath, '.js'),
+    log = defaultLog,
+  ) => {
+    const found = loaded.get(targetName);
+    // console.log('load', { targetName, found: !!found, rootPath });
+    if (found && found.rootPath === rootPath) {
+      return found.bundle;
+    }
+    const todo = makePromiseKit();
+    loaded.set(targetName, { rootPath, bundle: todo.promise });
+    const bundle = await validateOrAdd(rootPath, targetName, log)
+      .then(({ bundleFileName }) =>
+        import(`${wr.readOnly().neighbor(bundleFileName)}`),
+      )
+      .then(m => harden(m.default));
+    assert.equal(bundle.moduleFormat, 'endoZipBase64');
+    todo.resolve(bundle);
+    return bundle;
+  };
+
+  return harden({
+    add,
+    validate,
+    validateOrAdd,
+    load,
+  });
+};
+
+/**
+ * @param {string} dest
+ * @param {{ format?: string, dev?: boolean }} options
+ * @param {(id: string) => Promise<any>} loadModule
+ */
+export const makeNodeBundleCache = async (dest, options, loadModule) => {
+  const [fs, path, url, crypto, timers] = await Promise.all([
+    await loadModule('fs'),
+    await loadModule('path'),
+    await loadModule('url'),
+    await loadModule('crypto'),
+    await loadModule('timers'),
+  ]);
+
+  const readPowers = {
+    ...makeReadPowers({ fs, url, crypto }),
+    delay: ms => new Promise(resolve => timers.setTimeout(resolve, ms)),
+    basename: path.basename,
+  };
+
+  const cwd = makeFileReader('', { fs, path });
+  await fs.promises.mkdir(dest, { recursive: true });
+  const destWr = makeFileWriter(dest, { fs, path });
+  return makeBundleCache(destWr, cwd, readPowers, options);
+};
+
+/**
+ * @param {[string, string, string[]]} args
+ * @param {object} powers
+ * @param {(spec: string) => any} powers.loadModule
+ * @returns {void}
+ */
+export const main = async (args, { loadModule }) => {
```

Comment:
command line handling shouldn't be a concern of this module. please move into `src/index.js`